gem/optimise.py

```diff
@@ -413,3 +474,758 @@ def aggressive_unroll(expression):
+        """
+        Constructor
+        :param node: gem node
+        :param arg_ind: tuples of tuples of argument (linear) indices
+        """
+        self.node = node
+        self.arg_ind = arg_ind
+        # TODO: make these two properties of the object
+        self.arg_ind_flat = tuple([i for id in self.arg_ind for i in id])
+        self.arg_ind_set = set(self.arg_ind_flat)
+        if isinstance(node, IndexSum):
+            self.multiindex = node.multiindex
+            self.node = node.children[0]
+        else:
+            self.multiindex = ()
+        if rep:
+            self.rep = rep
+        else:
+            self._build_repr()
+
+
+    def _decide_key(self, factor):
+        """
+        Helper function to decide the appropriate key of a factor
+        :param factor: gem node
+        """
+        fi = factor.free_indices
+        if not fi:
+            return 'const'
+        else:
+            ind_set = set(fi) & arg_ind_set
+            if len(ind_set) > 1:
+                return tuple(i for i in arg_ind_flat if i in ind_set)
+            elif len(ind_set) == 0:
+                return 'quad'
+            else:
+                return ind_set.pop()
+
+    def _build_repr(self):
+        """
+        Build self.repr from self.node
+        """
+        node = expand_products(self.node, self.arg_ind_flat)
+        summands = collect_terms(node, Sum)
+        rep = []
+        for summand in summands:
+            d = OrderedDict()
+            for i in ['const', 'quad'] + list(self.arg_ind_flat):
+                d[i] = list()
+            for factor in collect_terms(summand, Product):
+                key = self._decide_key(factor)
+                if key in d:
+                    d[key].append(factor)
+                else:
+                    d[key] = [factor]
+            rep.append(d)
+        self.rep = rep
+
+    def find_optimal_arg_factors(self):
+        # TODO: add guard so that self.rep is up to date here
+        rep = self.rep
+        arg_ind_flat = tuple([i for id in self.arg_ind for i in id])
+        arg_ind_set = set(self.arg_ind_flat)
+
+        gem_int = OrderedDict()  # Gem node -> int
+        int_gem = OrderedDict()  # int -> Gem node
+        extent = dict()  # stores product of extents of free indices
+        counter = 0
+        # TODO: perhaps should keep a list of all nodes in the object
+        # assign number to all factors
+        for summand in rep:
+            for key, factors in summand.items():
+                # TODO: this pattern appears multiple times, need to rewrite
+                if key == 'const' or key == 'quad':
+                    continue
+                for factor in factors:
+                    if factor not in gem_int:
+                        gem_int[factor] = counter
+                        int_gem[counter] = factor
+                        extent[counter] = numpy.product([i.extent for i in set(factor.free_indices) & arg_ind_set])
```

Comment:
Make extent a local function, so you're not carrying around an extra dictionary:

```python
def extent(factor):
    return numpy.product([i.extent for i in factor.free_indices if i in argument_indices])
```